### PR TITLE
feat: request new tokens before they expire

### LINKED ISSUE #230: feat: consider moving eager refresh logic into gtoken from auth
@stephenplusplus suggested it might be worth considering moving eager refresh logic from the auth library into gtoken itself, let's investigate this at some point and see what a hassle it would be.

https://github.com/googleapis/node-gtoken/pull/229

### LINKED ISSUE #348: Can return tokens that will expire before the request is successfully sent
Because new tokens are not fetched until the old one has expired, it is possible to have some requests fail when they happen right in the moment when a token expires. I suggest we allow a leeway of 60 seconds for the actual request to be prepared, sent and authenticated. Given that the tokens are valid for 60 minutes, we will still be making use of the vast majority of the time that a token is valid for.

### PR DESCRIPTION
**Pull request by ForbesLindesay**
Thank you for opening a Pull Request! Before submitting your PR, there are a few things you can do to make sure it goes smoothly:
- [x] Make sure to open an issue as a [bug/issue](https://github.com/googleapis/node-gtoken/issues/new/choose) before writing your code!  That way we can discuss the change, evaluate designs, and agree on the general idea
- [x] Ensure the tests and linter pass
- [x] Code coverage does not decrease (if any source code was changed)
- [x] Appropriate docs were updated (if necessary)

Fixes #230 🦕

### CODE REVIEW

**Comment by ForbesLindesay** [comment]
@googlebot I signed it!

File: test/index.ts [hunk 1]
```diff
@@ -336,6 +361,18 @@ describe('.getToken()', () => {
...
 
+  it('should not use cached token if expiring soon', async () => {
+    const gtoken = new GoogleToken(TESTDATA);
+    gtoken.rawToken = {
+      access_token: 'mytoken',
+    };
+    gtoken.expiresAt = new Date().getTime() + 4 * 60 * 1000;
+    const fakeToken = 'abc123';
+    const scope = createGetTokenMock(200, {access_token: fakeToken});
+    const token = await gtoken.getToken({forceRefresh: true});
```
Comment (bcoe) [inline test/index.ts:372]: If possible, let's get rid of some of the warnings. Can you call `scope.done()` to ensure that the mock finished, as an example?

File: test/index.ts [hunk 2]
```diff
@@ -138,6 +138,31 @@ describe('.hasExpired()', () => {
...
 });
 
+describe('.isTokenExpiring()', () => {
+  it('should exist', () => {
+    const gtoken = new GoogleToken();
+    assert.strictEqual(typeof gtoken.isTokenExpiring, 'function');
+  });
+
+  it('should detect expiring tokens', () => {
+    const gtoken = new GoogleToken();
```
Comment (bcoe) [inline test/index.ts:148]: For ease of migrating this behavior from the auth library to the token library, I think we should make the functionality disabled by default.

**Review by bcoe — CHANGES_REQUESTED** [review]
This seems like a solid addition to the library, I think we might want to make the eagerRefresh behavior opt in though (with a test that makes sure it's not applied by default).

This will make it easier to move the auth library over to this version of gtoken.

**Comment by ForbesLindesay** [comment]
I've restored the original default behaviour of 0 ms for eager refresh. I've tried to keep my change fairly minimal to make it easy to merge, but I did add the `scope` assertions as requested.

File: test/index.ts [hunk 3]
```diff
@@ -138,6 +138,33 @@ describe('.hasExpired()', () => {
   });
 });
 
+describe('.isTokenExpiring()', () => {
+  it('should exist', () => {
+    const gtoken = new GoogleToken();
+    assert.strictEqual(typeof gtoken.isTokenExpiring, 'function');
```
Comment (bcoe) [inline test/index.ts:144]: I think we should add a test for the default behavior of `eagerRefreshThresholdMillis = 0`, which I believe should always indicate `isTokenExpiring()` of `false?

**Review by bcoe — COMMENTED** [review]
👍 this is looking good to me; I might be missing it, but I think one test for the default behavior of not refreshing would be good.

Once we release this, we can look at changing the behavior of `google-auth-library`. Out of curiosity, are you using gtoken directly, or using it through the auth library?

**Comment by ForbesLindesay** [comment]
> I might be missing it, but I think one test for the default behavior of not refreshing would be good.

The original tests already cover this case. The library always refreshed expired tokens.

See "should return cached token if not expired"

**Comment by ForbesLindesay** [comment]
We're using it directly. The layers on top just added masses of complexity that were extremely difficult to dig through when all we needed was a token to use in a fetch request.

File: test/index.ts (hunk 3 above)
Comment (ForbesLindesay) [reply test/index.ts:144]: I've added a test for this

**Comment by bcoe** [comment]
@ForbesLindesay not quite sure why CI/CD is failing after the last update. If we can get things green, let's land this.

**Review by bcoe — APPROVED** [review]

**Review by bcoe — APPROVED** [review]

**Comment by bcoe** [comment]
@ForbesLindesay thank you for the contribution 👍 

This is now out the door in `5.2.0`, let me know if it's working well for you?

**Pull request merged**